Conditional wait for non WebElement entities

dbudim opened the pull request:
## Proposed changes
Added conditional wait for non WebElement entities like LocalStorage. Now we can wait for some condition by Predicate.

## Checklist
- [x] Checkstyle and unit tests are passed locally with my changes by running `gradlew check chrome_headless firefox_headless` command
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

rosolko commented on src/main/java/com/codeborne/selenide/ConditionWaiter.java:
```diff
@@ -0,0 +1,42 @@
+  public static void waitFor(Conditional conditional, Predicate predicate, String message, int timeout) {
+    try {
+      await()
+        .pollInterval(1, SECONDS)
```
In any way this case should be configurable.
You can reload method with default value, but add an ability to set this parameter with custom value as well as `timeout`.

rosolko commented on src/main/java/com/codeborne/selenide/ConditionWaiter.java:
```diff
@@ -0,0 +1,42 @@
+        .pollInSameThread()
+        .until(() -> checkUnThrowable(conditional, predicate));
+    } catch (ConditionTimeoutException e) {
+      throw new ConditionTimeoutException(message);
```
Re-wrap this exception with selenide one.

dbudim replied:
The most appropriate that i found was SoftAssertionError, i used it. 
But i think we can create new on like ConditionNotMatchException)

BorisOsipov replied:
yes we have to create new exception like it is for element conditions.

dbudim replied:
aded

rosolko commented on src/main/java/com/codeborne/selenide/ConditionWaiter.java:
```diff
@@ -0,0 +1,42 @@
+    try {
+      return predicate.test(conditional);
+    } catch (Exception e) {
+      logger.info("Fail to check condition." + System.lineSeparator() + e);
```
```suggestion
      logger.info("Fail to check condition", e);
```

rosolko commented on src/main/java/com/codeborne/selenide/Conditional.java:
```diff
@@ -6,6 +6,6 @@
   void should(Predicate<T> predicate, String message);
 
-  void should(Predicate<T> predicate, String message, int timeout);
+  void should(Predicate<T> predicate, String message, long timeout);
```
I don't think that `int`/`long` meet selenide api.
For example methods with any sort of wait such `shouldBe` or `shouldBe` using `Duration` type of `timeout`.
Also we already deprecate methods `waitUntil` or `waitWhile` with `long` type `timeout` parameter.
Is is possible to use `Duration` instead of `long` in awaility library methods

rosolko commented on src/main/java/com/codeborne/selenide/Conditional.java:
```diff
@@ -0,0 +1,12 @@
+
+  void should(Predicate<T> predicate, String message);
+
+  void should(Predicate<T> predicate, String message, Duration timeout);
```
Look like you're missing condition with `polling`

dbudim replied:
I thought that polling is not so frequent configurable but if you say, i'll do it

rosolko replied:
There is no sense to make ability to configure pooling and not extend  interface with it

rosolko commented on src/main/java/com/codeborne/selenide/LocalStorage.java:
```diff
@@ -1,10 +1,24 @@
+  @Override
+  public void should(Predicate<LocalStorage> predicate, String message, Duration timeout) {
+    ConditionWaiter.waitFor(this, predicate, message, timeout);
+  }
```
ditto

rosolko commented on src/main/java/com/codeborne/selenide/LocalStorageConditions.java:
```diff
@@ -0,0 +1,15 @@
+    return localStorage -> localStorage.getItem(item).equals(value);
+  }
+
+}
```
ditto

dbudim replied:
Here is no polling, only predicate. What are you mean?

rosolko replied:
Opps, miss comment

BorisOsipov commented:
@dbudim Thank you for help with this feature. That's awesome.

I have only one concern - you've added `awatility` deps for this.
We already have the same retry feature in https://github.com/selenide/selenide/blob/master/src/main/java/com/codeborne/selenide/impl/SelenideElementProxy.java#L130 with simple `do..while`. What do you think about to avoid adding `awatility` deps and make it works with our own retry impl?

BorisOsipov commented:
>While starting work on this PR i've asked about this moment here and we agreed that awaitility can be used so i've used it)

Sure you can. My post is not about to restrict\allow smth. It is my opinion :)

BorisOsipov commented:
One thing we forget. Will we take screen\pagesource on fail?

Users say that is a must - https://github.com/selenide/selenide/issues/280#issuecomment-858523685

dbudim commented:
> One thing we forget. Will we take screen\pagesource on fail?
> 
> Users say that is a must - [#280 (comment)](https://github.com/selenide/selenide/issues/280#issuecomment-858523685)

I've checked) Yes)
![image](https://user-images.githubusercontent.com/46810751/121524091-9f37a680-c9ff-11eb-9fcf-773de5f3d334.png)

BorisOsipov commented:
>I've checked) Yes)

nope) it is screenshots from https://github.com/selenide/selenide/blob/f9553715609644a74f99c6643afe91371219ed8f/statics/src/test/java/integration/IntegrationTest.java#L34 junit extension.

Also it will be nice to embed screen\source path to exception as we do for element condition

dbudim commented:
> > I've checked) Yes)
> 
> nope) it is screenshots from
> 
> https://github.com/selenide/selenide/blob/f9553715609644a74f99c6643afe91371219ed8f/statics/src/test/java/integration/IntegrationTest.java#L34
> 
> junit extension.
> Also it will be nice to embed screen\source path to exception as we do for element condition

You mean 

> > I've checked) Yes)
> 
> nope) it is screenshots from
> 
> https://github.com/selenide/selenide/blob/f9553715609644a74f99c6643afe91371219ed8f/statics/src/test/java/integration/IntegrationTest.java#L34
> 
> junit extension.
> Also it will be nice to embed screen\source path to exception as we do for element condition

Am i on right way?) 
I used in `UIAssertionError.wrap` as like in `ElementsCollection`)
Now Exception looks like:
![image](https://user-images.githubusercontent.com/46810751/121582881-3a4c7280-ca38-11eb-8bcb-10dd7f904a8d.png)

But for it we need to put Driver into ConditionWaiter, so now usage will be `new ConditionWaiter(driver).waitFor();`


Is it ok?)

BorisOsipov commented:
>Am i on right way?)
>I used in UIAssertionError.wrap as like in ElementsCollection)

I guess yes

bereg2k commented on src/main/java/com/codeborne/selenide/LocalStorageConditions.java:
```diff
@@ -0,0 +1,15 @@
+
+public class LocalStorageConditions {
+
+  public static Predicate<LocalStorage> itemPresent(String item) {
```
For the sake of readability, it'd better rename it to `containItem()` or `item()` so that the user's code would read like this:

`localStorage().should(containItem("mouse"),"Item 'mouse' doesn't exist")`

`localStorage().shouldHave(item("mouse"),"Item 'mouse' doesn't exist")`

dbudim replied:
you right)
renamed itemPresent -> containItem

bereg2k replied:
`containItem()`. 

`containsItem()` reads grammatically incorrect if you consider it as a part of *"should containS item"*.

dbudim replied:
sure?
what about `containsKey` in Map?

dbudim replied:
You right, i followed `containsKey` in `Map` convention, but in case of `should` it looks better)
done

bereg2k commented on src/main/java/com/codeborne/selenide/LocalStorageConditions.java:
```diff
@@ -0,0 +1,15 @@
+    return localStorage -> localStorage.containsItem(item);
+  }
+
+  public static Predicate<LocalStorage> itemHasValue(String item, String value) {
```
For the sake of readability, it'd better rename it to `itemWithValue()` so that the user's code would read like this:

`localStorage().shouldHave(itemWithValue("mouse", "Jerry"),"Item 'mouse' value doesn't match")`

dbudim replied:
not sure. We have only `should(predicate)` method method,` so should(itemWithValue)` looks not pretty)

bereg2k replied:
If we can only utilize "should", then I'm suggesting `should(containItemWithValue(...))` or something like that. It will also look good next to the suggested above `should(containItem(...))`.

dbudim replied:
Thanks) looks nice) 
done)

bereg2k commented on src/main/java/com/codeborne/selenide/ConditionWaiter.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Class provide conditional polling wait by predicate with ignore exceptions.
+ */
+
```
is it ok that there's an empty line between JavaDoc and the class's declaration?

dbudim replied:
removed blank line

dbudim commented:
@asolntsev @dstekanov Waiting for your review)

asolntsev reviewed (changes requested):
1. It seems a better English name for `ConditionNotMatchException` is `ConditionNotMetException` or `ConditionNotFulfilled`. See https://stackoverflow.com/questions/63751794/what-does-the-condition-is-met-exactly-mean-in-programming-languages
2. We still have hard-coded polling interval 500ms in `ConditionWaiter`. Should take it from `driver.config().pollingInterval()`.
3. Do we really need `Awaitility` dependency in `ConditionWaiter`? It seems to be a simple do..while loop.
4. I have found Selenide already has class very similar to `ConditionWaiter`: `com.codeborne.selenide.impl.Waiter`. 
  * It doesn't catch exceptions. I think we can add try/catch to `Waiter`.

dbudim commented on src/main/java/com/codeborne/selenide/ConditionWaiter.java:
```diff
@@ -0,0 +1,46 @@
+    } catch (ConditionTimeoutException e) {
+      throw new SoftAssertionError(message);
+    }
+  }
```
We have discussed it few times before)
I think that awaitility is more convenient and smooth rather than making own "bicycle" with do while)

BorisOsipov replied:
@dbudim 

Dmitriy let's remove awaitlity. At least two mainteiners have concerns about it. Simple solution like `do..while` or `com.codeborne.selenide.impl.Waiter` are better than new dependency :)

dbudim replied:
Okay) 
We can improve `Waiter` class, as i see it just waits and doesn't throw any exception?
should we add separate method which will be throw if condition not met until timeout? Or we can modify existed method?

asolntsev replied:
It seems we can modify the existing method. I am surprised it doesn't throw an exception. :)

dbudim commented on src/main/java/com/codeborne/selenide/Conditional.java:
```diff
@@ -0,0 +1,14 @@
+import java.time.Duration;
+import java.util.function.Predicate;
+
+public interface Conditional<T> {
```
Not sure)
As for many programmers choosing good name is challenging for me)

What about `Awaitable` or `CanWaitCondition` ?)

asolntsev replied:
Then Conditional is better. :)
One more option is `Subject`, but I am not sure if it's better.

dbudim commented:
> 1. It seems a better English name for `ConditionNotMatchException` is `ConditionNotMetException` or `ConditionNotFulfilled`. See https://stackoverflow.com/questions/63751794/what-does-the-condition-is-met-exactly-mean-in-programming-languages
agree, will make

> 2. We still have hard-coded polling interval 500ms in `ConditionWaiter`. Should take it from `driver.config().pollingInterval()`.
agree will add

> 3. Do we really need `Awaitility` dependency in `ConditionWaiter`? It seems to be a simple do..while loop.
I prefer to use Awaitility it more convenient and smooth ) but if you say i'll make it with do while

> 4. I have found Selenide already has class very similar to `ConditionWaiter`: `com.codeborne.selenide.impl.Waiter`.
> 
> * It doesn't catch exceptions. I think we can add try/catch to `Waiter`.
Do you mean add to existed methods or add waitUnthrowable?

BorisOsipov commented:
@dbudim another thing I realized today we forgot about events for reporters.
Commands and asserts execution produces events that reporter like allure uses to enrich in report.

for example, for element it done here
https://github.com/selenide/selenide/blob/b8b91aea105a754e611111fab144ed3f5d436568/src/main/java/com/codeborne/selenide/impl/SelenideElementProxy.java#L83

it will be nice if it works for not element asserts too.

dbudim commented:
> @dbudim another thing I realized today we forgot about events for reporters.
> Commands and asserts execution produces events that reporter like allure uses to enrich in report.
> 
> for example, for element it done here
> https://github.com/selenide/selenide/blob/b8b91aea105a754e611111fab144ed3f5d436568/src/main/java/com/codeborne/selenide/impl/SelenideElementProxy.java#L83
> 
> it will be nice if it works for not element asserts too.

You right)
Agreed with @asolntsev that he will make it)

dbudim commented:
@asolntsev @BorisOsipov 
I removed awaitility))

Used `Waiter` class, but for getting default values from config and making screenshots it is needed to put `Driver` into `Waiter`.
I checked usages and didn't see ability to get driver in that places, so i decided do not change constructor and just create separate `wait` methods which puts Driver as argument.
Also i modified check predicate for unThrowable check)

Please let me know what you think about it)

rosolko reviewed (approved):
Looks pretty!
